src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/FunctionPointerCallingConventionClassifier.cs

```diff
@@ -0,0 +1,43 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using Microsoft.CodeAnalysis.Classification;
+using Microsoft.CodeAnalysis.Classification.Classifiers;
+using Microsoft.CodeAnalysis.CSharp.LanguageServices;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.CSharp.Classification
+{
+    internal class FunctionPointerCallingConventionClassifier : AbstractSyntaxClassifier
+    {
+        public override ImmutableArray<Type> SyntaxNodeTypes { get; } = ImmutableArray.Create(
+            typeof(FunctionPointerUnmanagedCallingConventionSyntax));
+
+        public override void AddClassifications(
+            SyntaxNode syntax,
+            SemanticModel semanticModel,
+            ClassificationOptions options,
+            ArrayBuilder<ClassifiedSpan> result,
+            CancellationToken cancellationToken)
+        {
+            var callingConventionSyntax = (FunctionPointerUnmanagedCallingConventionSyntax)syntax;
+            if (CSharpSyntaxFacts.Instance.IsSpecialUnmanagedCallingConvention(callingConventionSyntax))
+            {
+                result.Add(new ClassifiedSpan(callingConventionSyntax.Span, ClassificationTypeNames.ClassName));
```

Comment:
As mentioned on Discord, this might actually be more appropriate to be highlighted as a keyword rather than a class, given the compiler is not looking up any type, and there's no associated type symbol here. As discussed with Fred, these identifier here are effectively "a sort of contextual keyword". Related, we might want to update the coloring for this case in the tooltip as well.

Reply:
@CyrusNajmabadi What do you think about classifying as a keyword?

Also I've been looking about the logic that handles function pointers in quick info but can't find where it's. Can you point me out?

Reply:
i am fine with this being classified as a class

Reply:
Isn't that incorrect here? As per our conversation with Fred yesterday (who said he'd be fine with these being classified as a keyword), those are more akin to keywords than classes, given the compiler does not look the types at all. If we classified those as classes, we'd both (1) have an inconsistency in that no symbol info would be available (eg. in quick info), and (2) not actually mirror what the compiler is doing at all, as it's not actually using or looking up those types at all. According to the ECMA spec those calling conventions here are just special symbols, not classes.

Reply:
> Isn't that incorrect here?

This is classification.  There is no concept of 'correctness' :)

> According to the ECMA spec those calling conventions here are just special symbols, not classes.

That's fine.  I'm still ok with it being classified in this fashion.  Classification is not hte compiler.  It's whatever we want it to be.

Reply:
I understand that, but what I'm saying is that given those special symbols are effectively more akin to contextual keywords here (I mean, they basically are) than classes, and the fact that no type symbol would be available at all for them (which is not what you'd expect when hovering on a token being classified as a class), wouldn't classifying them as keywords be less confusing and more appropriate in this scenario? I can understand that classification can be arbitrary, but I really feel like classifying them as keywords would make more sense and be more consistent 🙂

Reply:
@Sergio0694 fwiw, the symbol display (part of the compiler) was classifying them as class names :)

Here is how everything looks with the latest changes.

https://user-images.githubusercontent.com/31348972/151191986-f36f8c55-9be5-4b9d-b87b-03fdd8541d60.mp4

Reply:
Imho, this is absolutely perfect, I **love** it! Amazing work! 😍

Reply:
> I understand that, but what I'm saying is that given those special symbols are effectively more akin to contextual keywords here (I mean, they basically are) than classes

Honestly, from teh perspective of classification, i don't think it matters.  That they are special cases isn't really meaningful.    We have all the CallConv types.  Sure, it's encoded one way and not referenced, but consistency with those seems fine to me.

Reply:
The point is that, other than reflecting what the ECMA says and the compiler does, using keywords also explains why there is no type info available there, which would otherwise arguably just look broken. You would never see a token being classified as a class and with no type info being available on it. I'd surely report it as a bug if I ever saw such a case.

Reply:
I genuinely think it does not matter.  And i would accept that behavior :)

Reply:
I guess what I'm saying is, if you just think it doesn't matter (which I read as you not being strongly opposed to it either), would you be fine approving this with those tokens being classified as keywords then? I genuinely think this behavior would be better, and Fred also mentioned he'd be fine with those tokens being classified as keywords in this case.

Reply:
> We have all the CallConv types.

You might not have all the CallConv types.

Reply:
What i mean is that for the purposes of this feature we know those always succeed, and can just presume they're fully legal all the time.

Reply:
> What i mean is that for the purposes of this feature we know those always succeed, and can just presume they're fully legal all the time.

We know that those _words_ always succeed. The types might literally not exist, depending on the TFM.

Reply:
> We know that those words always succeed. The types might literally not exist, depending on the TFM.

that's fine.  i'm still ok classifying it just like SymbolDisplay does.  It seems totally reasonable to me for the purposes of classification.  I don't have a pressing belief that these should display otherwise.

Reply:
@CyrusNajmabadi One other aspect I see is completion.

Completion will fallback to showing them as "keyword" if the types didn't exist. It definitely looks weird if I picked a keyword from completion and find it being classified as a class name. So that's a point of inconsistency to consider.

Reply:
my caring level about the inconsistency there is super low.  i'd also be totally fine with completion showing them as classes too. :)

Reply:
stepping back, IDE goals are not aligned with a view that every single thing we show or do must exactly match the language and/or be 100% in sync with a compiler API here.

As an example, i just opened https://github.com/dotnet/roslyn/pull/59086 where we will just unilaterally show `var` in a type-context as the `var-keyword` even if a user has a type named `var` in scope.

Why?  Because we just do not care, nor want extra work being done to have the experience light up for that niche customer wanting to squat on things in this fashion.

The language certainly has to consider things like back compat.  However, the IDE can absolutely take a position of "yeah, it's just reasonable to do things uniformly here, having subtle distinctions just isn't worth it".

Reply:
My feeling is that:

- completion, quick info, and classification should all always be consistent.
- we shouldn't classify something as a class without "go to definition" being working, and it cannot work if the types don't exist.
- we shouldn't classify something as a class if the compiler doesn't care about the existence of this class.

If classifying as keyword isn't too convincing, I'd leave it unclassified rather than classifying as class name.

Reply:
I will say I am honestly quite confused about this. Cyrus keeps saying it's "not important", which I mean, fine, but why block this? Youssef and I both feel like the behavior in the current PR is both more consistent and generally "better", and Fred also said he'd be fine with classifying those types as keywords. I don't understand why you're blocking it. I can understand saying it's not important and not considering this a priority, but given we already have a community PR implementing this in a way that'd arguably great, I just don't get what the problem is with this.

Reply:
i'm saying i don't think it's important to have a distinction between the special types and the other types.  And i'd be fine with goto-def taking you to the metadata for those calling convs, even if it's not precisely what is referenced in the emitted code.

> but given we already have a community PR implementing this in a way that'd arguably great, I just don't get what the problem is with this.

I'd prefer it be done differently.

Reply:
@CyrusNajmabadi If I'll do it in a different way, I'll classify the special ones as "identifiers" (or something else other than classes) instead of keywords.

One thing to keep in mind is what compiler behavior for the **public method** `ToDisplayParts` should be. That's an overlap between the IDE and the compiler because quick info relies on the compiler behavior (the change I made to `SymbolDisplayVisitor`).

Reply:
> "I'd prefer it be done differently"

I get that, but what you're suggesting just seems worse to me for all the reasons Yussef, Fred and I mentioned. You said that's fine and that it's not important, which again I can understand, but what I honestly don't understand is what the point is of _blocking_ this until the worse solution is done instead. I would understand if this had been something that you or others from the team had to implement, in which case not caring about this and not wanting to spend time on it would've been perfectly understandable, but given Yussef has already done the work I really don't get why we're getting stuck on this. Fred has already mentioned that, as per spec, those types don't exist, and that he personally feels like classifying those special identifiers as keywords would be fine. So does Yussef and so do I, and the current PR already implements all of this just fine, it works and it also looks great. Isn't that... Fine? 😅

Reply:
My approach requires zero compiler compiler changes and makes the ide clarification consistent with the compiler. I feel like that's a good thing.  It's not a"worse solution" to me. That is strictly better as:

1. It requires no compiler changes
2. It's is much simpler and cheaper on the ide side. 
3. It keeps compiler and ide classification in sync. 

You keep attacking me as of you think I'm saying let's do a worse solution muahahaha.  I ask that you please recognize that this may be a case where there are different views on the subject and that the things you find important may not be important to me at all.

Reply:
> given Yussef has already done the work I really don't get why we're getting stuck on this. 

Regardless of who does this, I still want the right experience to go in, and I want to consider things like complexity and maintenance. To be, I would just treat all of these conventions uniformly.  I think that's an entirely acceptable experience (same as how the compiler layer does that for symbol display).

Reply:
> "You keep attacking me as if you think I'm saying let's do a worse solution"

It was absolutely not my intention to attack you (I mean why would I ever want to do that Cyrus?), so I'm really sorry if my messages came across that way. I was just genuinely confused about why the proposed changes were being an issue for you, is all. I do recognize that you just had a different view on the topic, and I was actively trying to understand that better. Nothing of what I said was ever meant to be an attack towards you at all, same as with all other times we happened to disagree on things before as well, of course. I do understand your point better now that you've listed those specific aspects related to complexity and maintainability 🙂

Reply:
Sorry, 'attack' was too strong a word.  I likely wasn't making my thoughts very clear (i tend to be juggling like 15 things at at a time :)).  

> If this point is essential, then I think the key point to get to an agreement is to first define what's the expected output of the compiler API (especially that it has effect on the public ToDisplayParts method).

Yes, i agree with that.  If the compiler goes that way, i'm ok keepign IDE in sync.

Reply:
I'll leave you decide which approach is the best, just happy we could clarify the misunderstanding here ❤️

Reply:
The compiler will not give a type for a single work of `Stdcall`, `Thiscall`, `Cdecl`, or `Fastcall`. Regardless of whether we decide to return symbols for other combinations of those those words or other calling conventions, we specifically _do not_ look at symbols for those cases. They do not correspond to types and are not emitted as modreqs on the function pointer.